yarpl/test/FlowableFlatMapTest.cpp

```diff
@@ -0,0 +1,277 @@
+      in, [=](int64_t i) { return (i >= startat) && (i < endat); });
+}
+
+auto make_flowable_mapper_func() {
+  return folly::Function<Reference<Flowable<int64_t>>(int)>([](int n) {
+    switch (n) {
+      case 10:
+        return Flowables::range(n, 2);
+      case 20:
+        return Flowables::range(n, 3);
+      case 30:
+        return Flowables::range(n, 4);
+    }
+    return Flowables::range(n, 3);
+  });
+}
+
+// assumes that separate streams of values in separate_streams are entirely
+// disjoint
+template <typename T>
+bool validate_flatmapped_values(
+    std::vector<T> flatmapped,
+    std::vector<std::deque<T>> separate_streams) {
+  for (auto elem : flatmapped) {
+    bool found_match = false;
+    for (auto& stream : separate_streams) {
+      if (stream.size() > 0) {
+        if (elem == stream[0]) {
+          stream.pop_front();
+          found_match = true;
+          break;
+        }
+      }
+    }
+
+    EXPECT_TRUE(found_match)
+        << "Did not find elem '" << elem << "' in any input streams";
+    if (!found_match) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+TEST(FlowableFlatMapTest, AllRequestedTest) {
+  auto f =
+      Flowables::justN<int>({10, 20, 30})->flatMap(make_flowable_mapper_func());
+
+  std::vector<int64_t> res = run(f);
+  EXPECT_EQ(9UL, res.size());
+  EXPECT_EQ(filter_range(res, 10, 20), std::vector<int64_t>({10, 11}));
+  EXPECT_EQ(filter_range(res, 20, 30), std::vector<int64_t>({20, 21, 22}));
+  EXPECT_EQ(filter_range(res, 30, 40), std::vector<int64_t>({30, 31, 32, 33}));
+}
+
+TEST(FlowableFlatMapTest, FiniteRequested) {
+  auto f =
+      Flowables::justN<int>({10, 20, 30})->flatMap(make_flowable_mapper_func());
+
+  auto subscriber = make_ref<TestSubscriber<int64_t>>(1);
+  f->subscribe(subscriber);
+
+  EXPECT_EQ(1UL, subscriber->values().size());
+  EXPECT_TRUE(
+      validate_flatmapped_values(subscriber->values(), {{10}, {20}, {30}}));
+
+  subscriber->request(3);
+  EXPECT_TRUE(validate_flatmapped_values(
+      subscriber->values(), {{10, 11}, {20, 21, 22}, {30, 31, 32, 33}}));
+
+  subscriber->cancel();
+  EXPECT_EQ(subscriber->getValueCount(), 4);
+}
+
+TEST(FlowableFlatMapTest, MappingLambdaThrowsErrorOnFirstCall) {
+  folly::Function<Reference<Flowable<int64_t>>(int)> func = [](int n) {
+    CHECK_EQ(1, n);
+    throw std::runtime_error{"throwing in mapper!"};
+    return Flowables::empty<int64_t>();
+  };
+
+  auto f = Flowables::just<int>(1)->flatMap(std::move(func));
+
+  auto subscriber = make_ref<TestSubscriber<int64_t>>(1);
+  f->subscribe(subscriber);
+
+  EXPECT_EQ(subscriber->getValueCount(), 0);
+  EXPECT_TRUE(subscriber->isError());
+  EXPECT_EQ(subscriber->getErrorMsg(), "throwing in mapper!");
+}
+
+TEST(FlowableFlatMapTest, MappedStreamThrows) {
+  folly::Function<Reference<Flowable<int64_t>>(int)> func = [](int n) {
+    CHECK_EQ(1, n);
+
+    // flowable which emits an onNext, then the next iteration, emits an error
+    int64_t i = 1;
```

Comment:
this 'i' will get deleted before it gets used?, already you pass it by value, so how can it become 0 at all?

Reply:
Because it's captured by value in a mutable lambda, so it'll dec each time the lambda fires